src/Evolution/DgSubcell/GhostZoneLogicalCoordinates.hpp

```diff
@@ -0,0 +1,93 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <cstddef>
+
+#include "DataStructures/DataVector.hpp"
+#include "DataStructures/Index.hpp"
+#include "DataStructures/IndexIterator.hpp"
+#include "DataStructures/Tensor/Tensor.hpp"
+#include "Domain/Structure/Direction.hpp"
+#include "NumericalAlgorithms/Spectral/Mesh.hpp"
+#include "NumericalAlgorithms/Spectral/Spectral.hpp"
+#include "Utilities/ContainerHelpers.hpp"
+#include "Utilities/ErrorHandling/Assert.hpp"
+#include "Utilities/Gsl.hpp"
+
+namespace evolution::dg::subcell::fd {
+
+/*!
+ * \brief Compute logical coordinates of ghost points for given subcell mesh
+ * extents and a given direction.
+ *
+ */
+template <size_t Dim>
+void ghost_zone_logical_coordinates(
+    const gsl::not_null<tnsr::I<DataVector, Dim, Frame::ElementLogical>*>
+        ghost_logical_coords,
+    const Mesh<Dim>& subcell_mesh, const size_t ghost_zone_size,
+    const Direction<Dim>& direction) {
+  const auto subcell_extents = subcell_mesh.extents();
+  const size_t dim_direction = direction.dimension();
+  const size_t subcell_extent_to_direction = subcell_extents[dim_direction];
+
+  // Check if the `ghost_zone_size` has a valid value.
+  ASSERT(ghost_zone_size <= subcell_extent_to_direction,
+         " Ghost zone size ("
+             << ghost_zone_size << ") is larger than the volume extent ("
+             << subcell_extent_to_direction << ") to the direction");
+
+  destructive_resize_components(
+      ghost_logical_coords,
+      ghost_zone_size *
+          subcell_mesh.extents().slice_away(dim_direction).product());
+
+  Index<Dim> ghost_zone_extents{};
+  for (size_t d = 0; d < Dim; ++d) {
+    ghost_zone_extents[d] =
+        d == dim_direction ? ghost_zone_size : subcell_extents[d];
+  }
+
+  const auto get_fd_collocation_points = [](const size_t num_fd_pts) {
+    return Spectral::collocation_points<Spectral::Basis::FiniteDifference,
+                                        Spectral::Quadrature::CellCentered>(
+        num_fd_pts);
+  };
+
+  for (size_t d = 0; d < Dim; ++d) {
+    const auto& collocation_points_in_this_dim =
+        get_fd_collocation_points(subcell_extents[d]);
+    const double delta_x{collocation_points_in_this_dim[1] -
+                         collocation_points_in_this_dim[0]};
+
+    for (IndexIterator<Dim> index(ghost_zone_extents); index; ++index) {
+      if (d == dim_direction) {
```

Comment:
I would move the `if` outside the for loop. No point in doing a for loop with a conditional check every iteration. I would even factor out the conditional `direction.side() == Side::Upper` and use a `size_t index_offset = direction.side() == Side::Upper ? ... : 0;`. Avoiding branching in for loops is good for performance

Reply:
Okay, I'll move all branching out of index `for` loop